index.rst

```diff
@@ -10,16 +10,200 @@
+Once pipeline processing is complete, the worker uploads its outputs to the central repository.
+The worker will likely also compute metrics at this stage, but the details (in particular, whether they are computed before or after the central repository sync) have not yet been decided.
+
+For performance reasons, each worker places pipeline outputs from all visits into a small number of Butler runs, which are persisted when each worker uploads to the central repository.
+The run name is a deterministic function of the Science Pipelines configuration, the instrument and choice of pipeline, and the processing date.
+The use of a separate run for each pipeline is required by the Gen 3 provenance system (in particular, the storage of pipeline configuration information and "init-output" datasets), but makes it difficult for the Prompt Processing system to adjust its choice of pipeline in response to missing inputs or to processing failures.
+
+.. _general-priorities:
+
+General Priorities
+==================
+
+Of the outputs produced by the Alert Production pipeline, the most important to have correct and uncorrupted is the APDB, which serves as the source of truth both for later processing and for the externally available Prompt Products database.
+The alert stream is a close second, since it is the primary user interface for transient follow-up and cannot be "fixed" after the fact.
+The central repository and metrics are less important, because they are (almost) exclusively for internal use.
+
+As far as possible, the Prompt Processing pipeline should be idempotent -- in other words, multiple tries of the same pipeline on the same dataset should yield the same results, no matter what the state of the system.
+This is possible when rebroadcasting the alert stream -- in the worst case, duplicate alerts exist but have the same position and time -- and when running the pipeline from ISR through image differencing.
+However, for reasons explored in :ref:`association`, idempotency cannot be safely achieved with source association.
+
+.. _retries:
+
+Retrying Processing
+===================
+
+The execution framework can automatically recover from failures, in the sense of restoring full processing capacity without compromising the framework's state.
+However, recovering the data processing itself is the responsibility of the code running on each worker.
+
+Visit processing can fail at any point, including within and between tasks, and while performing external I/O.
+Even within a single worker, the order of events is not fully deterministic (e.g., two snaps could be ingested and processed up to snap combination in parallel).
+However, the possible failures can be divided into a handful of cases:
+
+- Any failure before the end of APDB writing (which is atomic as of May 2023) has no consequences beyond that worker's local repository.
```

Comment:
It's good (necessary even) to point out the atomicity of APDB updates, but it might be good to also insert a forward reference here to the difficulties of making them idempotent.